sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/SharedRntbdTransportClient.java

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.directconnectivity;
+
+import com.azure.cosmos.implementation.Configs;
+import com.azure.cosmos.implementation.LifeCycleUtils;
+import com.azure.cosmos.implementation.RxDocumentServiceRequest;
+import com.azure.cosmos.implementation.UserAgentContainer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This class uses a shared RntbdTransportClient for multiple Cosmos Clients.
+ * The benefit is the underlying connections can be shared if possible across multiple Cosmos client instances.
+ */
+public class SharedRntbdTransportClient extends TransportClient {
+    private static final Logger logger = LoggerFactory.getLogger(SharedRntbdTransportClient.class);
+    private static final AtomicInteger counter = new AtomicInteger(0);
+    private static SharedRntbdTransportClient sharedTransportClient;
+
+    public static TransportClient getOrCreateInstance(Configs configs, int requestTimeout, UserAgentContainer userAgent) {
+        synchronized (TransportClient.class) {
+            if (sharedTransportClient == null) {
+                assert counter.get() == 0;
+                logger.info("creating a new shared RntbdTransportClient");
+                sharedTransportClient = new SharedRntbdTransportClient(configs, requestTimeout, userAgent);
+            } else {
+                logger.info("Reusing an instance of RntbdTransportClient");
+            }
+
+            counter.incrementAndGet();
+            return sharedTransportClient;
+        }
+    }
+
+    private final TransportClient transportClient;
+
+    private SharedRntbdTransportClient(Configs configs, int requestTimeout, UserAgentContainer userAgent) {
+        this.transportClient = new RntbdTransportClient(configs, requestTimeout, userAgent);
+    }
+
+    @Override
+    protected Mono<StoreResponse> invokeStoreAsync(Uri physicalAddress, RxDocumentServiceRequest request) {
+        return transportClient.invokeStoreAsync(physicalAddress, request);
+    }
+
+    public int getReferenceCounter() {
+        return counter.get();
+    }
+
+    @Override
+    public void close() throws Exception {
+        synchronized (TransportClient.class) {
```

Comment:
What is the reason of synchronizing on TransportClient and not on SharedRntbdTransportClient, isn't the threads entering this block on the instance of SharedRntbdTransportClient ?

Reply:
good question, here we are not locking on an instance of class, this lock is a global JVM lock on the class object rather than an instance. So doing it on either on `SharedRntbdTransportClient.class` or `TransportClient.class` both are JVM global lock.

Reply:
change that to the `SharedTransportClient.class` to be cleaner in the code.